samples/android/tutorial-4-opencl/jni/GLrender.cpp

```diff
@@ -0,0 +1,315 @@
+        0.0f, 0.0f,
+        1.0f, 1.0f,
+        1.0f, 0.0f
+};
+float texCoord2D[] = {
+        0.0f, 0.0f,
+        0.0f, 1.0f,
+        1.0f, 0.0f,
+        1.0f, 1.0f
+};
+
+const char vss[] = \
+            "attribute vec2 vPosition;\n" \
+            "attribute vec2 vTexCoord;\n" \
+            "varying vec2 texCoord;\n" \
+            "void main() {\n" \
+            "  texCoord = vTexCoord;\n" \
+            "  gl_Position = vec4 ( vPosition.x, vPosition.y, 0.0, 1.0 );\n" \
+            "}";
+
+const char fssOES[] = \
+            "#extension GL_OES_EGL_image_external : require\n" \
+            "precision mediump float;\n" \
+            "uniform samplerExternalOES sTexture;\n" \
+            "varying vec2 texCoord;\n" \
+            "void main() {\n" \
+            "  gl_FragColor = texture2D(sTexture,texCoord);\n" \
+            "}";
+
+const char fss2D[] = \
+            "precision mediump float;\n" \
+            "uniform sampler2D sTexture;\n" \
+            "varying vec2 texCoord;\n" \
+            "void main() {\n" \
+            "  gl_FragColor = texture2D(sTexture,texCoord);\n" \
+            "}";
+
+int progOES = 0;
+int prog2D = 0;
+
+GLuint FBOtex = 0, FBOtex2 = 0;
+GLuint FBO = 0;
+
+GLuint texOES = 0;
+int texWidth = 0, texHeight = 0;
+
+static inline void deleteTex(GLuint* tex)
+{
+    if(tex && *tex)
+    {
+        glDeleteTextures(1, tex);
+        *tex = 0;
+    }
+}
+
+static void releaseFBO()
+{
+    if (FBO != 0)
+    {
+        glBindFramebuffer(GL_FRAMEBUFFER, 0);
+        glDeleteFramebuffers(1, &FBO);
+        FBO = 0;
+    }
+    deleteTex(&FBOtex);
+    deleteTex(&FBOtex2);
+    glDeleteProgram(prog2D);
+    prog2D = 0;
+}
+
+static inline void logShaderCompileError(GLuint shader)
+{
+    GLchar msg[512];
+    msg[0] = 0;
+    GLsizei len;
+    glGetShaderInfoLog(shader, sizeof(msg) - 1, &len, msg);
+    LOGE("Could not compile shader: %s", msg);
+}
+
+static int makeShaderProg(const char* vss, const char* fss)
+{
+    LOGD("makeShaderProg: setup GL_VERTEX_SHADER");
+    GLuint vshader = glCreateShader(GL_VERTEX_SHADER);
+    const GLchar* text = vss;
+    glShaderSource(vshader, 1, &text, 0);
+    glCompileShader(vshader);
+    int compiled;
+    glGetShaderiv(vshader, GL_COMPILE_STATUS, &compiled);
+    if (compiled == 0) {
+        logShaderCompileError(vshader);
+        glDeleteShader(vshader);
+        vshader = 0;
+    }
+
+    LOGD("makeShaderProg: setup GL_FRAGMENT_SHADER");
+    GLuint fshader = glCreateShader(GL_FRAGMENT_SHADER);
+    text = fss;
+    glShaderSource(fshader, 1, &text, 0);
+    glCompileShader(fshader);
+    glGetShaderiv(fshader, GL_COMPILE_STATUS, &compiled);
+    if (compiled == 0) {
+        logShaderCompileError(fshader);
+        glDeleteShader(fshader);
+        fshader = 0;
+    }
+
+    LOGD("makeShaderProg: glCreateProgram");
+    GLuint program = glCreateProgram();
+    glAttachShader(program, vshader);
+    glAttachShader(program, fshader);
+    glLinkProgram(program);
```

Comment:
you should also check `GL_LINK_STATUS` via `glGetProgramiv`, then call `glValidateProgram` and check `GL_VALIDATE_STATUS` too.